lib/lf/geometry/test/geometry_tests.cc

```diff
@@ -98,6 +98,43 @@ void checkIntegrationElement(const lf::geometry::Geometry &geom,
   }
 }
 
+/**
+ * Checks that SubGeometry and Geometry map the same nodes to the same points
+ */
+void checkSubGeometry(const lf::geometry::Geometry &geom) {
+  // nodeCoords is a (refEl.Dimension, refEl.NumNodes) matrix
+  const auto refEl = geom.RefEl();
+  const Eigen::MatrixXd &nodeCoords = refEl.NodeCoords();
+
+  // iterate over all relative codimensions
+  for (size_t codim = 1; codim <= geom.RefEl().Dimension(); ++codim) {
+    // iterate over all subEntities in given codimension
+    const auto numSubEntities = refEl.NumSubEntities(codim);
+    for (size_t subEntity = 0; subEntity < numSubEntities; ++subEntity) {
+      // subNodeCoords is a (subRefEl.Dimension, subRefEl.NumNodes) matrix
+      auto subGeom = geom.SubGeometry(codim, subEntity);
+      auto subRefEl = subGeom->RefEl();
+      const Eigen::MatrixXd &subNodeCoords = subRefEl.NodeCoords();
+
+      // iterate over all nodes of subEntity
+      for (size_t subNode = 0; subNode < subRefEl.NumNodes(); ++subNode) {
```

Comment:
I think this is already a quite good test, congrats. The only problem is a bit that it checks the property `SubGeometry().Global(x) == Global(\xi(x))` only at the corner nodes of the sub-reference element. For second order elements this is not really enough because the mapping maybe correct at the endpoints but it could be that the middle point is somehow wrongly set:
![grafik](https://user-images.githubusercontent.com/15138672/48092449-015b1a00-e20d-11e8-82ea-83f8c41004fe.png)

a more thorough test idea would be to add an additional parameter to the `checkSubGeometry()` method: 
```
void checkSubGeometry(const Geometry& geom, std::function<RefEl,QuadRule> qr_provider)
```
with the idea that the property `SubGeometry().Global(x) == Global(\xi(x))` is tested for all `x` in `qr_provider(subRefEl).points()`. Note that in order to write such a test you can construct the mapping \xi using a first order Geometry object that maps from `subRefEl` to the corners `nodeCoords.col(subSubIdx)` (using the variable names that you have used in your code).

Reply:
But wouldn't `TriaO2::NodeCoords()` return the coordinates of the middle point too?

Reply:
I'm a bit confused: I don't think that the `lf::geometry::Geometry` interface class has a member `TriaO2::NodeCoords()`? Maybe you're thinking about the reference element `TriaO2`? In the latter case I want to point out that there is no need to introduce a new reference element! The reference element of the `TriaO2` geometry element is also `lf::base::RefEl::kTria()`...

Reply:
The reference element for any cell with three edges (no matter whether straight or curved) and three vertices is the "unit triangle" `kTria()`. All special features of the geometry have to be introduced by means of special mappings encoded in the `rf::geometry` classes.

Please think of the reference element mainly as a representative of the _topological class_ of an entity.